File: solutions/c/bob/1/bob.c
#include "bob.h"

#include<string.h>
#include<ctype.h>

bool is_question(char* input){
    int count = strlen(input)-1;
    bool q = false;    
    while(count>=0 && isspace(input[count])){
        count--;
    }
    if(count>=0) q = q || input[count]=='?';
    return q;      
}

bool is_shouting(char* input){
    bool caps = true;
    bool has_letters = false;
    for(size_t i=0; i<strlen(input); i++){        
        caps = isalpha(input[i]) ? caps && isupper(input[i]) : caps;
        if(!has_letters && isalpha(input[i])) has_letters = true;
    }
    return has_letters && caps;
}

bool is_silence(char* input){
    bool silence = true;
    for(size_t i=0; i<strlen(input); i++){
        silence = silence && isspace(input[i]);
    } 
    return silence;
}

char *hey_bob(char *greeting){
    char *res;
    if (is_silence(greeting)){
        res = "Fine. Be that way!";} 
    else if (is_question(greeting)) {
        if(is_shouting(greeting)) { res = "Calm down, I know what I'm doing!"; }
        else { res = "Sure.";} }
    else if (is_shouting(greeting)){ 
        res="Whoa, chill out!";}
    else { res="Whatever.";}
   return res;
}

File: solutions/dart/bob/1/lib/bob.dart
enum Intent{Silence, Question, Yelling, YellQuestion, AllElse}

class Bob {
  static final List<String> responses = [
        "Sure.", "Whoa, chill out!", "Calm down, I know what I'm doing!",  "Fine. Be that way!", "Whatever." ];

  bool is_question(String input) {
         String str = input.replaceAll(
           RegExp(r'\s+'),'');
        return str[str.length-1] == '?';
    }

   bool is_yelling(String input){
        final lower = RegExp(r'[a-z]');
        final upper = RegExp(r'[A-Z]');
     
        return !lower.hasMatch(input) 
          && upper.hasMatch(input);
   }

  bool is_silence(String input){
    return !RegExp(r'\S').hasMatch(input);
  }

  Intent classify(String input){
    if(is_silence(input))
      return Intent.Silence;
    var isQ = is_question(input);
    var isY = is_yelling(input);
    if(isQ && isY)
      return Intent.YellQuestion;
    else if(isQ)
      return Intent.Question;
    else if(isY)
      return Intent.Yelling;
    else
      return Intent.AllElse;    
  }

  String response(String input){
    String res = '';
    switch(classify(input)){
      case Intent.Silence : res = 'Fine. Be that way!';
      case Intent.Question : res = 'Sure.';
      case Intent.YellQuestion : res = "Calm down, I know what I'm doing!";
      case Intent.Yelling : res = 'Whoa, chill out!';
      default : res = 'Whatever.';
   }
    return res;
  }
  
}
